### ISSUE TITLE
Spring Boot dashboard live information shows up in a separate PID 

### ISSUE THREAD

**Issue by nickzhums** [microsoft/vscode-spring-boot-dashboard#248]
Repro steps:
1 Clone a spring petclinic project
2 Run project from Spring Boot dashboard

Java extension pack and spring boot dashboard all using latest version

Both Tanya and I can repro this from the machine. Interesting thing is that this problem only happens for one of my machine (consistently) but not other machine. 

Screenshot for the problem

![MicrosoftTeams-image (3)](https://user-images.githubusercontent.com/56864335/202135293-719b22ce-4aee-42dd-96ac-ee6775f59526.png)

https://user-images.githubusercontent.com/56864335/202193963-5b1785f6-0733-40a2-87be-9161c14d698b.mp4

**Comment by nickzhums** [comment]
thank you, I have been having this issue for a long time on my machine, would appreciate a fix. Given that tangtang has this issue as well, other users might encounter the same problem

**Comment by Eskibear** [comment]
> other users might encounter the same problem

@nickzhums yes, for sure. but I myself cannot reproduce it. If we can find a way to stably reproduce it, that would be much helpful.

**Comment by Eskibear** [comment]
On the same macbook pro, I can reproduce it with code-insiders, but not able to repro it with code. Let's compare the differences.

project: (I removed `.git` and `target` folders, hope it won't affect the result)
[spring-petclinic.zip](https://github.com/microsoft/vscode-spring-boot-dashboard/files/10072622/spring-petclinic.zip)


#### On code-insiders (can repro)

Command: 
```
/usr/bin/env /Users/yanzh/.sdkman/candidates/java/17.0.5-ms/bin/java @/var/folders/wd/kwf6ls4n6dq7nw8vw9rc03tr0000gp/T/cp_nxu7pmodyegq5mrc9h41xn9u.argfile org.springframework.samples.petclinic.PetClinicApplication 
```

[argfile.txt](https://github.com/microsoft/vscode-spring-boot-dashboard/files/10072579/argfile.txt)

#### On code (cannot repro)
command: 
```
/usr/bin/env /Users/yanzh/.sdkman/candidates/java/11.0.16.1-ms/bin/java @/var/folders/wd/kwf6ls4n6dq7nw8vw9rc03tr0000gp/T/cp_a1lz01tvw7mxr0o9fehhxwpud.argfile org.springframework.samples.petclinic.PetClinicApplication
```

[argfile2.txt](https://github.com/microsoft/vscode-spring-boot-dashboard/files/10072590/argfile2.txt)


For both, `sun.java.command` has the correct value. 

One problem is, in pom.xml `<java.version>11</java.version>` is defined. Why in code-insiders it use java 17 to launch the project? @testforstephen  any idea?

**Comment by nickzhums** [comment]
Do you think it might be caused by JDK ? I 'm also using JDK 17 on my machine

**Comment by Eskibear** [comment]
It can be. After adding below settings, forcing to use a JDK 11 to launch the project, the issue is gone.
```
{
    "java.configuration.runtimes": [
    {
        "name": "JavaSE-11",
        "path": "/Users/yanzh/.sdkman/candidates/java/11.0.16.1-ms"
    }
    ]
}
```


But I don't know the root cause.

**Comment by nickzhums** [comment]
I see, still better to find out what the root cause is. Could also leverage Martin if needed

**Comment by nickzhums** [comment]
Okay I'm having a weird issue. I downloaded JDK 11. I modified my POM to be JDK 11. I changed my settings.json and set it to JDK 11. 

Somehow in Java Project explorer, even though it says JDK 11, the path still points to JDK 17's path. 

When I click play button, it still uses JDK 17 to start up the project, therefore, still re-producing the same problem. Here's the video.


https://user-images.githubusercontent.com/56864335/203469681-7e295925-e116-489f-9bd7-b64a1212c5c1.mp4

**Comment by Eskibear** [comment]
<img width="1121" alt="image" src="https://user-images.githubusercontent.com/2351748/203470823-9eec4ed1-c015-4086-afaf-ab916778c6c0.png">

The processName "Unknown" is from here.

**Comment by martinlippert** [comment]
@Eskibear Strange... I am trying to reproduce this locally. I have `petclinic` with Java level set to 11 in my workspace and run it using a JDK 17, but the process name shows up nicely. So something must be different here on my side that is causing this issue to not appear.

The exact command line that appears in the output view when running petclinic here is:
`/usr/bin/env /usr/bin/java -agentlib:jdwp=transport=dt_socket,server=n,suspend=
y,address=localhost:52585 -XX:+ShowCodeDetailsInExceptionMessages -Dspring.jmx.enabled=true -Dmanagement.endpoints.jmx.exposure.include=\* -Dspring.application.admin.enabled=true
 -Dspring.boot.project.name=spring-petclinic -Djava.rmi.server.hostname=localhost @/var/folders/k5/nndcf2h95r56w1qnzp6sw9lr0000gn/T/cp_ef0908rwwtemar991urayshx6.argfile org.sprin
gframework.samples.petclinic.PetClinicApplication`

`/usr/bin/java` points to:

`OpenJDK 64-Bit Server VM Temurin-17.0.4+8 (build 17.0.4+8, mixed mode, sharing)`

**Comment by Eskibear** [comment]
I'm thinking whether it's because the distribution of JDK 17. So I installed and switch to `17.0.4.1-tem`, and the issue is gone. 
Then I switched back to use `17.0.5-ms`, the issue is also gone. Now I'm cannot reproduce it any more. It's really strange, not sure what's happening. I assume it's something related to the impl about how `VirtualMachineDescriptor` is initialized (or is the displayname modified somewhere in sts?),  or perhaps we have to look into the JDK itself...

**Comment by martinlippert** [comment]
@Eskibear I am not sure what exactly is causing this, but the `Unknown` process name comes from a different source. When the actual live data connector (the one via JMX in this case) was initialized with no process name, it tries to infer a meaningful process name (inside of `SpringProcessLiveDataExtractorOverJMX`) via the system properties of the running process, directly asking for the java command property.

It feels to me this is overly complex and goes through too many loops and abstraction layers. Why do you really need the process name? It sounds like you are trying to extract a main class name from it to identify the corresponding project in the workspace? Or the corresponding launch config?

**Comment by Eskibear** [comment]
> Why do you really need the process name? It sounds like you are trying to extract a main class name from it to identify the corresponding project in the workspace?

It's a good question letting us get back to talk about the real requirements. Yes, I use it to identify corresponding project and update the status in views (apps,beans,mappings). The project has 3 states like, not-running, launching (spinning icon), running. When you run your project, it switches to "launching" state and will change to "running" state when corresponding live process is connected. That's the time we think its service is ready and you can access it. Back to the time I implemented it, dashboard extension might have no knowledge about pid of the running debug session depending on where it's launched (in internal console or terminal). That's why we use main class name to match project and live process.

Even if the "Unknown" issue got fixed, I still feel not comfortable to have the assumption that a processName is the main class, without any contract to ensure it. Now since [pid is reported by java debugger extension](https://github.com/microsoft/java-debug/pull/413#issuecomment-1140652571), it should be possible to use `pid` to link local projects and the live processes, and I think that should be the right way to go.

**Comment by Eskibear** [comment]
Closing as completed. #256 implemented debugger-reported PID matching for the live process.
